documentation/adr/0018-algorithm-migration-to-paged-files.md

```diff
@@ -0,0 +1,120 @@
+[<-previous](0017-separate-absolute-and-relative-indexing-APIs-in-pixel-array.md) | next->
+
+# 18 - Algorithm migration to paged implementations
+
+Date: 2020-Oct-15
+
+## Status
+
+Accepted
+
+## Context
+
+The codebase includes two different `cut()` methods:
+ - one in the SQW object API which operates on the in-memory SQW data
+ - one is a function that takes a `.sqw` filename and operates directly on that (on-disk)
+
+The on-disk implementation is significantly slower and users are aware of this.
+If a 'too large' file is operated on in-memory, an OutOfMemory Error will be thrown and the process will fail.
+A typical use-case for `cut` consists of taking an initial cut from a large dataset held in a file
+and storing this cut in memory as an `sqw` or `dnd` object.
+Subsequent cuts are in-memory object to the in-memory object.
+
+As a consequence:
+ - users are required to develop (and test) separate scripts for each API
+ - users make a conscious choice between “in memory (fast)” and “on disk (slow)” operations.
+
+Mantid have (multiple times) attempted to present seamless in-memory/on-disk operation to their users (for a slightly different use case).
+Each time this either failed for technical reasons or because users did not embrace it,
+and the attempts were abandoned.
```

Comment:
Small modification:
Mantid have successfully implemented paging operations in two situations: Paged instrument workspace, corresponding to workspace, generated for single experiment and MDWorkspace, functionally equivalent to Horace SQW object. Unfortunately, the success was only "technical" and in fact was the failure, at least for ISIS, as the code is not used by the target scientific community and, in the first case, has been fully removed from the code base due to parlous performance of file based operations with respect to memory operations. In the second case the reasons for failure are more subtle but analysis of the users complains points to:
 - negative user experience due to substantial performance drop at implicit change from memory-based to file-based  operations
 - steep learning curve, as the users, concerning about their scientific task are requested to specify deeply technical information substantially affecting performance (e.g. page size, and/or thread count) from the beginning. 
 - unclear and unusual interface, including the need to perform the requested operations in two steps, i.e. creating object and doing the operations over then.  Despite the later sound very minor, the number of users, who finds this way of operating the package confusing is surprising. 

Horace workflow, from other side, remains natural an popular way of analyzing the results of experiments, so should be kept as much as reasonable.  In addition to that, the idea that if one invokes file operation it may run slowly, and if it is memory operation it should be fast is very natural to any our users.